File: frontend_server/routes/webflow.js
import express from 'express'
import { assertEnvVar } from '../utils.js'

const router = express.Router()

/** * Configuration Constants
 * Using the CDN endpoint for improved performance and higher rate limits.
 */
const WEBFLOW_API_TOKEN = assertEnvVar('WEBFLOW_API_TOKEN')
const WEBFLOW_NEWS_COLLECTION_ID = '6811154eaf44a4127de5c768'
const WEBFLOW_TAGS_COLLECTION_ID = '681260687cb93bd6bf4f782f'
const HET_ORG_ID = 'ae155e605db764fb19ba82c7a5b4ac26'
const WEBFLOW_BASE = 'https://api-cdn.webflow.com/v2'

const WEBFLOW_HEADERS = {
  Authorization: `Bearer ${WEBFLOW_API_TOKEN}`,
  accept: 'application/json',
}

/** * Cache State
 * TTL-based caching to minimize external API latency and stay within rate limits.
 */
let tagCache = null
let tagFetchTime = 0
let newsCache = null
let newsFetchTime = 0

const TAG_TTL = 60 * 60 * 1000  // 1 Hour
const NEWS_TTL = 5 * 60 * 1000  // 5 Minutes

/**
 * Retrieves tag metadata from Webflow.
 * Implements a stale-if-error pattern to ensure service availability.
 * @param {boolean} forceRefresh - If true, bypasses the local TTL cache.
 * @returns {Promise<Object>} Map of tag IDs to display names.
 */
async function getTagMap(forceRefresh = false) {
  const now = Date.now()
  if (!forceRefresh && tagCache && (now - tagFetchTime < TAG_TTL)) {
    return tagCache
  }

  try {
    const res = await fetch(`${WEBFLOW_BASE}/collections/${WEBFLOW_TAGS_COLLECTION_ID}/items`, {
      headers: WEBFLOW_HEADERS
    })

    if (!res.ok) throw new Error(`Tags API error: ${res.statusText}`)

    const data = await res.json()
    tagCache = Object.fromEntries(data.items.map((item) => [item.id, item.fieldData.name]))
    tagFetchTime = now
    return tagCache
  } catch (err) {
    // Return stale cache if available, otherwise propagate error
    if (tagCache) {
      console.warn('[webflow-tags] API unreachable, serving stale cache.')
      return tagCache
    }
    throw err
  }
}

/**
 * GET /het-news
 * Retrieves the latest 3 news articles filtered by organization.
 */
router.get('/het-news', async (req, res) => {
  const now = Date.now()

  // Serve from cache if within TTL
  if (newsCache && (now - newsFetchTime < NEWS_TTL)) {
    return res.json({ articles: newsCache })
  }

  try {
    // Parallel fetch of tag metadata and news items
    let [tagMap, itemsRes] = await Promise.all([
      getTagMap(),
      fetch(`${WEBFLOW_BASE}/collections/${WEBFLOW_NEWS_COLLECTION_ID}/items`, {
        headers: WEBFLOW_HEADERS,
      }),
    ])

    if (!itemsRes.ok) {
      throw new Error(`News API error: ${itemsRes.status} ${itemsRes.statusText}`)
    }

    const data = await itemsRes.json()
    const rawItems = data.items || []

    /**
     * Lazy Cache Invalidation:
     * If an item contains a tag ID missing from our cache, trigger an
     * immediate refresh of the tag metadata.
     */
    const hasUnknownTag = rawItems.some((item) =>
      (item.fieldData.tags ?? []).some((id) => !tagMap[id])
    )

    if (hasUnknownTag) {
      tagMap = await getTagMap(true)
    }

    // Process and format articles
    const articles = rawItems
      .filter((item) => item.fieldData['primary-org'] === HET_ORG_ID)
      .sort((a, b) => new Date(b.fieldData.date) - new Date(a.fieldData.date))
      .slice(0, 3)
      .map((item) => {
        const f = item.fieldData
        const imageUrl = f['main-image']?.url
        return {
          title: f.name,
          author: f.author ?? null,
          date: f.date,
          tags: (f.tags ?? []).map((id) => tagMap[id] ?? id),
          slug: f.slug,
          summary: f['post-summary'] ?? null,
          thumbnail: imageUrl ? imageUrl.replace(/(\.\w+)(\?.*)?$/, '-p-500$1$2') : null, // fetch 500px wide version
        }
      })

    // Update news cache
    newsCache = articles
    newsFetchTime = now

    res.json({ articles })
  } catch (err) {
    console.error('[webflow-news] Error processing request:', err)

    // Recovery attempt: serve last known good data
    if (newsCache) {
      console.info('[webflow-news] Serving stale news articles after error.')
      return res.json({ articles: newsCache })
    }

    res.status(500).json({ error: 'Internal server error while fetching news articles.' })
  }
})

export default router

File: server/webflow.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"sort"
	"sync"
	"time"
)

const (
	webflowBase            = "https://api-cdn.webflow.com/v2"
	webflowNewsCollectionID = "6811154eaf44a4127de5c768"
	webflowTagsCollectionID = "681260687cb93bd6bf4f782f"
	hetOrgID               = "ae155e605db764fb19ba82c7a5b4ac26"
	tagTTL                 = time.Hour
	newsTTL                = 5 * time.Minute
)

type webflowArticle struct {
	Title     string   `json:"title"`
	Author    *string  `json:"author"`
	Date      string   `json:"date"`
	Tags      []string `json:"tags"`
	Slug      string   `json:"slug"`
	Summary   *string  `json:"summary"`
	Thumbnail *string  `json:"thumbnail"`
}

type webflowState struct {
	mu            sync.RWMutex
	tagCache      map[string]string
	tagFetchTime  time.Time
	newsCache     []webflowArticle
	newsFetchTime time.Time
}

var wfState webflowState

var webflowClient = &http.Client{Timeout: 10 * time.Second}

func webflowHeaders() map[string]string {
	token := os.Getenv("WEBFLOW_API_TOKEN")
	return map[string]string{
		"Authorization": "Bearer " + token,
		"Accept":        "application/json",
	}
}

func webflowGet(url string) (*http.Response, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	for k, v := range webflowHeaders() {
		req.Header.Set(k, v)
	}
	return webflowClient.Do(req)
}

func getTagMap(forceRefresh bool) (map[string]string, error) {
	now := time.Now()

	wfState.mu.RLock()
	if !forceRefresh && wfState.tagCache != nil && now.Sub(wfState.tagFetchTime) < tagTTL {
		tags := wfState.tagCache
		wfState.mu.RUnlock()
		return tags, nil
	}
	stale := wfState.tagCache
	wfState.mu.RUnlock()

	resp, err := webflowGet(fmt.Sprintf("%s/collections/%s/items", webflowBase, webflowTagsCollectionID))
	if err != nil {
		if stale != nil {
			log.Print("[webflow-tags] API unreachable, serving stale cache.")
			return stale, nil
		}
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		if stale != nil {
			log.Printf("[webflow-tags] API returned %d, serving stale cache.", resp.StatusCode)
			return stale, nil
		}
		return nil, fmt.Errorf("tags API error: %s", resp.Status)
	}

	var data struct {
		Items []struct {
			ID        string `json:"id"`
			FieldData struct {
				Name string `json:"name"`
			} `json:"fieldData"`
		} `json:"items"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		if stale != nil {
			return stale, nil
		}
		return nil, err
	}

	tags := make(map[string]string, len(data.Items))
	for _, item := range data.Items {
		tags[item.ID] = item.FieldData.Name
	}

	wfState.mu.Lock()
	wfState.tagCache = tags
	wfState.tagFetchTime = now
	wfState.mu.Unlock()

	return tags, nil
}

func hetNewsHandler(w http.ResponseWriter, r *http.Request) {
	if os.Getenv("WEBFLOW_API_TOKEN") == "" {
		http.Error(w, "Webflow API token not configured", http.StatusServiceUnavailable)
		return
	}

	now := time.Now()

	wfState.mu.RLock()
	if wfState.newsCache != nil && now.Sub(wfState.newsFetchTime) < newsTTL {
		articles := wfState.newsCache
		wfState.mu.RUnlock()
		writeJSON(w, map[string]any{"articles": articles})
		return
	}
	staleNews := wfState.newsCache
	wfState.mu.RUnlock()

	tagMap, err := getTagMap(false)
	if err != nil {
		log.Printf("[webflow-news] tag fetch error: %v", err)
		if staleNews != nil {
			log.Print("[webflow-news] serving stale news after tag error.")
			writeJSON(w, map[string]any{"articles": staleNews})
			return
		}
		http.Error(w, "Internal server error while fetching news articles.", http.StatusInternalServerError)
		return
	}

	resp, err := webflowGet(fmt.Sprintf("%s/collections/%s/items", webflowBase, webflowNewsCollectionID))
	if err != nil {
		log.Printf("[webflow-news] news fetch error: %v", err)
		if staleNews != nil {
			log.Print("[webflow-news] serving stale news after fetch error.")
			writeJSON(w, map[string]any{"articles": staleNews})
			return
		}
		http.Error(w, "Internal server error while fetching news articles.", http.StatusInternalServerError)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		log.Printf("[webflow-news] API returned %d", resp.StatusCode)
		if staleNews != nil {
			writeJSON(w, map[string]any{"articles": staleNews})
			return
		}
		http.Error(w, "Internal server error while fetching news articles.", http.StatusInternalServerError)
		return
	}

	var data struct {
		Items []struct {
			FieldData struct {
				Name       string   `json:"name"`
				Author     *string  `json:"author"`
				Date       string   `json:"date"`
				Tags       []string `json:"tags"`
				Slug       string   `json:"slug"`
				PostSummary *string `json:"post-summary"`
				PrimaryOrg string   `json:"primary-org"`
				MainImage  *struct {
					URL string `json:"url"`
				} `json:"main-image"`
			} `json:"fieldData"`
		} `json:"items"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		log.Printf("[webflow-news] decode error: %v", err)
		if staleNews != nil {
			writeJSON(w, map[string]any{"articles": staleNews})
			return
		}
		http.Error(w, "Internal server error while fetching news articles.", http.StatusInternalServerError)
		return
	}

	// Lazy tag cache invalidation: refresh if we find an unknown tag ID
	hasUnknownTag := false
	for _, item := range data.Items {
		for _, id := range item.FieldData.Tags {
			if _, ok := tagMap[id]; !ok {
				hasUnknownTag = true
				break
			}
		}
		if hasUnknownTag {
			break
		}
	}
	if hasUnknownTag {
		if refreshed, err := getTagMap(true); err == nil {
			tagMap = refreshed
		}
	}

	// Filter, sort, slice, and map to output shape
	type rawItem = struct {
		FieldData struct {
			Name       string   `json:"name"`
			Author     *string  `json:"author"`
			Date       string   `json:"date"`
			Tags       []string `json:"tags"`
			Slug       string   `json:"slug"`
			PostSummary *string `json:"post-summary"`
			PrimaryOrg string   `json:"primary-org"`
			MainImage  *struct {
				URL string `json:"url"`
			} `json:"main-image"`
		} `json:"fieldData"`
	}
	filtered := make([]rawItem, 0)
	for _, item := range data.Items {
		if item.FieldData.PrimaryOrg == hetOrgID {
			filtered = append(filtered, item)
		}
	}
	sort.Slice(filtered, func(i, j int) bool {
		return filtered[i].FieldData.Date > filtered[j].FieldData.Date
	})
	if len(filtered) > 3 {
		filtered = filtered[:3]
	}

	articles := make([]webflowArticle, 0, len(filtered))
	for _, item := range filtered {
		f := item.FieldData
		tags := make([]string, 0, len(f.Tags))
		for _, id := range f.Tags {
			name := tagMap[id]
			if name == "" {
				name = id
			}
			tags = append(tags, name)
		}
		var thumbnail *string
		if f.MainImage != nil {
			thumbnail = &f.MainImage.URL
		}
		articles = append(articles, webflowArticle{
			Title:     f.Name,
			Author:    f.Author,
			Date:      f.Date,
			Tags:      tags,
			Slug:      f.Slug,
			Summary:   f.PostSummary,
			Thumbnail: thumbnail,
		})
	}

	wfState.mu.Lock()
	wfState.newsCache = articles
	wfState.newsFetchTime = now
	wfState.mu.Unlock()

	writeJSON(w, map[string]any{"articles": articles})
}
